keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1055 @@
+  apply a read/write deadline.
+- Replace the asynchronous mutating-request executor with one that 
+  executes synchronously, in the same goroutine as the caller.
+  
+We also need to maintain parity with logs/metrics that are produced
+by the `timeout` filter today for any request that times out.
+
+### User Stories (Optional)
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+##### Client Hanging Indefinitely
+
+With the `timeout` filter enabled today, it sends a `504 GatewayTimeout` response
+or an `error` to the client, if the request handler it is executing is blocking indefinitely.
+
+If the client does not enforce any timeout on its side, for example, not setting a
+global `Client` timeout, or not setting a deadline to the request `Context`; the client
+can hang forever waiting for a response from the server. This is certainly true for
+HTTP/1x where [net/http does not execute the request handler in a new goroutine](https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/server.go#L2031-L2039).
+
+> It's worth noting that by default `client-go` does not enforce any timeout globally.
+
+With the removal of the `timeout` filter, we need to ensure that a request handler that
+blocks on the server should never block a client that does not enforce timeout on its side.
+
+In fact, we can reproduce a "forever" hanging client with http/1x:
+```
+func TestClientHangingForeverWithHTTP1(t *testing.T) {
+	clientDoneCh, handlerDoneCh := make(chan struct{}), make(chan error, 1)
+	server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		defer close(handlerDoneCh)
+
+		ctrl := http.NewResponseController(w)
+		if err := ctrl.SetWriteDeadline(time.Now().Add(200 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetWriteDeadline, but got: %v", err)
+			return
+		}
+		if err := ctrl.SetReadDeadline(time.Now().Add(100 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetReadDeadline, but got: %v", err)
+			return
+		}
+
+		<-clientDoneCh
+	}))
+
+
+	defer server.Close()
+	server.StartTLS()
+
+	client := server.Client()
+	func() {
+		defer close(clientDoneCh)
+		client.Get(server.URL + "/foo")
+	}()
+
+	select {
+	case <-handlerDoneCh:
+	case <-time.After(wait.ForeverTestTimeout):
+		t.Errorf("expected the request handler to have terminated")
+	}
+}
+```
+Even though a read/write deadline is set, the client never receives any 
+response or error from the server. We have opened an [issue](https://github.com/golang/go/issues/65526)
+with the golang team to track this.
+
+Mitigation Steps:
+- a) If this is a bug, we can wait until the fix is in the net/http package.
+- b) Otherwise, we can enable this feature only for http/2.0 request(s), all
+  http/1x request(s) will continue to be served using the `Timeout` filter.
```

Comment:
I haven't scrolled down yet, but I'm hoping to see an integration test of some kind for this with client-go against kube-apiserver.

Reply:
yes, it's listed